### ISSUE TITLE
Zstd adds floating point use between 1.3.4 and 1.3.7

### ISSUE THREAD

**Issue by cemeyer** [facebook/zstd#1386]
Hi Yann et al,

My primary use case is embedding Zstd in the FreeBSD kernel.  Our kernel runs in a mode where FPU use is not allowed by default, as a way of reducing context-switching overhead.  That means most of the kernel does not support floating point operation.

Some of Zstd's floating point uses are marked as debugging routines (`ZSTD_fWeight`, `ZSTD_fCost`) and it would be trivial to `#ifndef _KERNEL` kill them.  But I'm more concerned about e.g. `COVER_ctx_init()`, which grows a double argument and some use of floating point math.  Also `ZDICT_optimizeTrainFromBuffer_cover`, and some of the new fastcover stuff.

I'd like to be able to easily build Zstd without floating point support without extensive local patching; something like a `-DNO_FLOATING_POINT` would be ideal.  The easiest approach might just disable some functionality; that would be totally fine for us.  We don't have any immediate need for advanced stuff like dictionary building in the kernel, for example.  It would be awesome if the full functionality could be available with fixed point integer math instead of floats, but I understand that is more work, and I can't volunteer to do it myself right now.

(Our use cases are really quite basic -- we just compress MBs to GBs of userspace core dumps or kernel core dumps using the memory pre-allocation APIs at a given compression level.  See https://github.com/freebsd/freebsd/blob/master/sys/kern/subr_compressor.c#L251 .)

Thanks,
Conrad

**Comment by terrelln** [comment]
I'll remove the doubles from the dictionary builder.

Is there a compiler option to disable floating point, so we can add a test to our CI? Otherwise we risk regressing in the future.

**Comment by cemeyer** [comment]
I don't know of a specific compiler option, but one possible hack is to have the 'float' and 'double' keywords introduce syntax errors via `-D` compiler flags.  E.g., in our bootloader we use:

```
# Spike the floating point interfaces
CFLAGS+=     -Ddouble=jagged-little-pill -Dfloat=floaty-mcfloatface
```

That breaks the build when the floating point keywords are added accidentally.

**Comment by cemeyer** [comment]
For what it's worth, the following trivial patch is sufficient to mask the uses I can see (at least the ones GCC complains about :grin:).  You could imagine replacing `#if 0` with `#ifndef _KERNEL` or `#ifndef NO_FLOATING_POINT`.  It only seems to be used for `DEBUGLOG()`, so perhaps even `#if DEBUGLEVEL >= 2`.

```
--- zstd/lib/compress/zstd_compress_internal.h	(revision 340424)
+++ zstd/lib/compress/zstd_compress_internal.h	(working copy)
@@ -691,6 +691,7 @@
 
 /* debug functions */
 
+#if 0
 MEM_STATIC double ZSTD_fWeight(U32 rawStat)
 {
     U32 const fp_accuracy = 8;
@@ -714,6 +715,7 @@
                 u, table[u], ZSTD_fWeight(sum) - ZSTD_fWeight(table[u]) );
     }
 }
+#endif
 
 #if defined (__cplusplus)
 }
--- zstd/lib/compress/zstd_opt.c	(revision 340424)
+++ zstd/lib/compress/zstd_opt.c	(working copy)
@@ -53,10 +53,12 @@
 }
 
 /* debugging function, @return price in bytes */
+#if 0
 MEM_STATIC double ZSTD_fCost(U32 price)
 {
     return (double)price / (BITCOST_MULTIPLIER*8);
 }
+#endif
 
 static void ZSTD_setBasePrices(optState_t* optPtr, int optLevel)
 {
@@ -852,8 +854,10 @@
                     for ( ; pos <= end ; pos++ ) {
                         U32 const matchPrice = ZSTD_getMatchPrice(offset, pos, optStatePtr, optLevel);
                         U32 const sequencePrice = literalsPrice + matchPrice;
+#if 0
                         DEBUGLOG(7, "rPos:%u => set initial price : %.2f",
                                     pos, ZSTD_fCost(sequencePrice));
+#endif
                         opt[pos].mlen = pos;
                         opt[pos].off = offset;
                         opt[pos].litlen = litlen;
@@ -879,9 +883,11 @@
                                 - ZSTD_litLengthPrice(litlen-1, optStatePtr, optLevel);
                 assert(price < 1000000000); /* overflow check */
                 if (price <= opt[cur].price) {
+#if 0
                     DEBUGLOG(7, "cPos:%zi==rPos:%u : better price (%.2f<=%.2f) using literal (ll==%u) (hist:%u,%u,%u)",
                                 inr-istart, cur, ZSTD_fCost(price), ZSTD_fCost(opt[cur].price), litlen,
                                 opt[cur-1].rep[0], opt[cur-1].rep[1], opt[cur-1].rep[2]);
+#endif
                     opt[cur].mlen = 0;
                     opt[cur].off = 0;
                     opt[cur].litlen = litlen;
@@ -888,9 +894,11 @@
                     opt[cur].price = price;
                     memcpy(opt[cur].rep, opt[cur-1].rep, sizeof(opt[cur].rep));
                 } else {
+#if 0
                     DEBUGLOG(7, "cPos:%zi==rPos:%u : literal would cost more (%.2f>%.2f) (hist:%u,%u,%u)",
                                 inr-istart, cur, ZSTD_fCost(price), ZSTD_fCost(opt[cur].price),
                                 opt[cur].rep[0], opt[cur].rep[1], opt[cur].rep[2]);
+#endif
                 }
             }
 
@@ -947,8 +955,10 @@
                         int const price = basePrice + ZSTD_getMatchPrice(offset, mlen, optStatePtr, optLevel);
 
                         if ((pos > last_pos) || (price < opt[pos].price)) {
+#if 0
                             DEBUGLOG(7, "rPos:%u (ml=%2u) => new better price (%.2f<%.2f)",
                                         pos, mlen, ZSTD_fCost(price), ZSTD_fCost(opt[pos].price));
+#endif
                             while (last_pos < pos) { opt[last_pos+1].price = ZSTD_MAX_PRICE; last_pos++; }   /* fill empty positions */
                             opt[pos].mlen = mlen;
                             opt[pos].off = offset;
@@ -957,8 +967,10 @@
                             ZSTD_STATIC_ASSERT(sizeof(opt[pos].rep) == sizeof(repHistory));
                             memcpy(opt[pos].rep, &repHistory, sizeof(repHistory));
                         } else {
+#if 0
                             DEBUGLOG(7, "rPos:%u (ml=%2u) => new price is worse (%.2f>=%.2f)",
                                         pos, mlen, ZSTD_fCost(price), ZSTD_fCost(opt[pos].price));
+#endif
                             if (optLevel==0) break;  /* early update abort; gets ~+10% speed for about -0.01 ratio loss */
                         }
             }   }   }
```

Thanks!

**Comment by Cyan4973** [comment]
I've followed the proposed patch to remove debug functions using floating point from non-debug builds.

**Comment by Cyan4973** [comment]
In `v1.3.8`, `zstd` main decoder/encoder does no longer uses floating point in non-debug builds (`DEBUGLEVEL < 2`). 
Floating points are still present in optional capabilities, such as dictionary builder.

**Comment by cemeyer** [comment]
Fixed in v1.3.8!  Thanks!
